### PR TITLE
Make getSigkeysInAccount consider "del"

### PR DESCRIPTION
**Pull request by Powersource**
Fixes https://github.com/staltz/ppppp-db/issues/22

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -409,11 +409,18 @@ function initDB(peer, config) {
...
+        const purpose = data.key?.purpose
+        if (purpose !== 'sig' && purpose !== 'shs-and-sig') continue
+        if (data.key.algorithm !== 'ed25519') continue
+        sigkeys.add(data.key.bytes)
+      } else if (data.action === 'del') {
+        const purpose = data.key?.purpose
+        if (purpose !== 'sig' && purpose !== 'shs-and-sig') continue
+        if (data.key.algorithm !== 'ed25519') continue
+        // TODO: uncomment to get test to hopefully pass
+        //sigkeys.delete(data.key.bytes)
```
Comment (Powersource) [inline lib/index.js:422]: do this later. first I want the test to fail in the right way

File: lib/msg-v4/validation.js [hunk 2]
```diff
@@ -332,7 +332,11 @@ function validate(msg, tangle, sigkeys, msgID, rootID) {
   if ((err = validateSigkey(msg))) return err
   if ((err = validateData(msg))) return err
 
-  if (tangle.type === 'feed' && isMoot(msg)) return // nothing else to check
+  try {
+    if (tangle.type === 'feed' && isMoot(msg)) return // nothing else to check
+  } catch (err) {
+    return err
+  }
```
Comment (Powersource) [inline lib/msg-v4/validation.js:339]: `tangle.type` is a function that sometimes (like in this test) throws. Changed it to this to be able to see the err causes

File: test/account-add.test.js [hunk 3]
```diff
@@ -206,4 +206,91 @@ test('account.add()', async (t) => {
...
+      await p(carol.db.add)(accountRec1.msg, account)
+      console.log(
+        'rec',
+        goodRec,
+        'rectangle',
+        goodRec.msg.metadata.tangles,
+        'moot',
+        postMootRec
+      )
+      await p(carol.db.add)(goodRec.msg, postMootRec.id)
```
Comment (Powersource) [inline test/account-add.test.js:285]:
```
% node test/account-add.test.js
▶ account.add()
  ✔ Basic usage (414.540033ms)
  ✔ keypair with no "add" powers cannot add (793.603039ms)
  ✔ publish with a key in the account (531.707646ms)
rec {
  id: 'B1Qt4gRdfsjbvN6KeYJeVdDTXAk9sheFWKSKD4n8gZv4',
  msg: {
    data: { text: 'potato' },
    metadata: {
      dataHash: 'CvsZPpuqfYLEBC3TzbVMdxh91YEjhXQFe1Jdg46Zpn8L',
      dataSize: 17,
      account: 'jyUJX6tSFW3kU3vDSs79RRTqTtUuiW97upGLVRA61XF',
      accountTips: [Array],
      tangles: [Object],
      domain: 'post',
      v: 4
    },
    sigkey: 'FuUujJeb6TP7UAhWHi5HZMFP8fTejgPhnNyrx9c4GtwF',
    sig: '3UADDvQQ3QfcfC3joGzXQJeoako4V5e2TgugyfSk1CXGjroarrx2TWBXwLKWxtWiz3jMuuGVFgGeMES5iumyKF8o'
  },
  received: 1710714317072
} rectangle {
  '6fvVUU9WvbvUWWzggnHVMFsnWfbMutUBUXdCAR9jCYY6': {
    depth: 1,
    prev: [ '6fvVUU9WvbvUWWzggnHVMFsnWfbMutUBUXdCAR9jCYY6' ]
  }
} moot {
  id: '6fvVUU9WvbvUWWzggnHVMFsnWfbMutUBUXdCAR9jCYY6',
  msg: {
    data: null,
    metadata: {
      dataHash: null,
      dataSize: 0,
      account: 'jyUJX6tSFW3kU3vDSs79RRTqTtUuiW97upGLVRA61XF',
      accountTips: null,
      tangles: {},
      domain: 'post',
      v: 4
    },
    sigkey: 'FuUujJeb6TP7UAhWHi5HZMFP8fTejgPhnNyrx9c4GtwF',
    sig: '2vGBfd4L9xkzUJtMu9d2RL6h7YsM8XN3GsVZVv4x4juC8MUdLv5o5kgDh13AzVCsTKdhNFHf88nJPEQBRLmKmbP9'
  },
  received: 1710714317071
}
  ✖ Can't publish with a key if the key has been del'd (21.997003ms)
    Error: add() failed to verify msg
        at add (/home/me/prj/pzp/pzp-db/lib/index.js:570:17)
        at hooked (/home/me/prj/pzp/pzp-db/node_modules/hoox/index.js:10:15)
        ... 6 lines matching cause stack trace ...
        at async TestContext.<anonymous> (/home/me/prj/pzp/pzp-db/test/account-add.test.js:210:3)
        at async Test.run (node:internal/test_runner/test:632:9) {
      [cause]: Error: Invalid msg
          at verifyRec (/home/me/prj/pzp/pzp-db/lib/index.js:480:14)
          at add (/home/me/prj/pzp/pzp-db/lib/index.js:569:16)
          ... 5 lines matching cause stack trace ...
          at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
          at async Test.run (node:internal/test_runner/test:632:9)
          at async TestContext.<anonymous> (/home/me/prj/pzp/pzp-db/test/account-add.test.js:210:3) {
        [cause]: Error: Tangle "6fvVUU9WvbvUWWzggnHVMFsnWfbMutUBUXdCAR9jCYY6" is missing root message
            at get type [as type] (/home/me/prj/pzp/pzp-db/lib/msg-v4/tangle.js:229:13)
            at Object.validate (/home/me/prj/pzp/pzp-db/lib/msg-v4/validation.js:336:16)
            at verifyRec (/home/me/prj/pzp/pzp-db/lib/index.js:479:22)
            at add (/home/me/prj/pzp/pzp-db/lib/index.js:569:16)
            at hooked (/home/me/prj/pzp/pzp-db/node_modules/hoox/index.js:10:15)
            at node:internal/util:441:7
            at new Promise (<anonymous>)
            at hooked (node:internal/util:427:12)
            at TestContext.<anonymous> (/home/me/prj/pzp/pzp-db/test/account-add.test.js:285:28)
            at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      }
    }

▶ account.add() (1772.304172ms)
```

Comment (Powersource) [reply test/account-add.test.js:285]: @staltz am i doing something obviously wrong here?

File: test/account-add.test.js [hunk 4]
```diff
@@ -206,4 +206,91 @@ test('account.add()', async (t) => {
...
+      const goodRec = await p(peer.db.feed.publish)({
+        account,
+        domain: 'post',
+        data: { text: 'potato' },
+        keypair: keypair2,
+      })
+      const postMootRec = peer.db.feed.findMoot(account, 'post')
+
+      const accountRoot = peer.db.get(account)
+      const delRec = await p(peer.db.feed.publish)({
```
[the remark this answers is not in the document]
Comment (Powersource) [reply test/account-add.test.js:246]: but the error i posted above throws before getting to `add(delRec)`

yeah i thought about making this function but tried to limit the scope of this PR

[the remark this answers is not in the document]
Comment (staltz) [reply test/account-add.test.js:246]: > but the error i posted above throws before getting to `add(delRec)`

At what point exactly does it throw?

> yeah i thought about making this function but tried to limit the scope of this PR

Yes, but we'll need db.account.del anyways, so maybe you can make two PRs. If you want to finish this PR first, then you'll have to create an account tangle msg manually, using `MsgV4.create()` not `db.feed.publish()`.

[the remark this answers is not in the document]
Comment (Powersource) [reply test/account-add.test.js:246]: it throws on line 285 (according to the stacktrace), the line i left this comment on

yeah maybe i'll try .create() and see how far i get

[the remark this answers is not in the document]
Comment (Powersource) [reply test/account-add.test.js:246]: it throws on line 285, the line i left my comment on

yeah maybe i'll try .create() and see how far i get

[the remark this answers is not in the document]
Comment (Powersource) [reply test/account-add.test.js:246]: yeah commenting the delRec stuff still makes it throw the same way

File: test/account-add.test.js (hunk 3 above)
Comment (staltz) [reply test/account-add.test.js:285]: Hmm, `carol` has to `add()` the moot too.

File: test/account-add.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Powersource) [reply test/account-add.test.js]: you have to manually add moots apparently

File: test/account-add.test.js [hunk 5]
```diff
@@ -206,4 +207,93 @@ test('account.add()', async (t) => {
...
+      const carol = createPeer({ keypair: keypair3, path: DIR })
+
+      await carol.db.loaded()
+
+      await p(carol.db.add)(accountMsg0, account)
+      await p(carol.db.add)(accountRec1.msg, account)
+      await p(carol.db.add)(postMootRec.msg, postMootRec.id)
+      await p(carol.db.add)(goodRec.msg, postMootRec.id)
+      await p(carol.db.add)(delMsg, account)
+      await assert.rejects(
```
Comment (Powersource) [inline test/account-add.test.js:290]: It doesn't reject for some reason

```
✖ Can't publish with a key if the key has been del'd (106.61869ms)
    AssertionError [ERR_ASSERTION]: Missing expected rejection: Adding msg with del'd keypair is supposed to fail
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async TestContext.<anonymous> (/home/me/prj/pzp/pzp-db/test/account-add.test.js:290:7)
        at async Test.run (node:internal/test_runner/test:632:9)
        at async TestContext.<anonymous> (/home/me/prj/pzp/pzp-db/test/account-add.test.js:211:3)
        at async Test.run (node:internal/test_runner/test:632:9)
        at async startSubtest (node:internal/test_runner/harness:214:3) {
      generatedMessage: false,
      code: 'ERR_ASSERTION',
      actual: undefined,
      expected: /add\(\) failed to verify msg/,
      operator: 'rejects'
    }
```

**Review by staltz — APPROVED** [review]
All correct. Feel free to merge

**Pull request merged**